packages/core/src/domain/error/trackRuntimeError.spec.ts

```diff
@@ -260,55 +274,68 @@ describe('instrumentOnError', () => {
   const ERROR_MESSAGE = 'foo'
 
-  beforeEach(() => {
+  const withInstrumentOnUnhandledRejection = async (callback: () => void) => {
     if (!('onunhandledrejection' in window)) {
       pending('onunhandledrejection not supported')
     }
 
-    callbackSpy = jasmine.createSpy()
-    onUnhandledrejectionSpy = setupOnUnhandledrejectionSpy()
-    ;({ stop: stopCollectingUnhandledError } = instrumentUnhandledRejection(callbackSpy))
-  })
-
-  afterEach(() => {
-    stopCollectingUnhandledError?.()
-  })
+    const onUnhandledRejectionSpy = spyOn(window as any, 'onunhandledrejection')
+    const callbackSpy = jasmine.createSpy()
+    const { stop } = instrumentUnhandledRejection(callbackSpy)
 
-  it('should call original unhandled rejection handler', () => {
-    window.onunhandledrejection!({
-      reason: new Error(ERROR_MESSAGE),
-    } as PromiseRejectionEvent)
+    try {
+      await invokeAndWaitForErrorHandlers(callback)
+      expect(onUnhandledRejectionSpy).toHaveBeenCalled()
+      return callbackSpy
+    } finally {
+      stop()
+    }
+  }
 
-    expect(onUnhandledrejectionSpy).toHaveBeenCalled()
+  it('should call original unhandled rejection handler', async () => {
+    // withInstrumentOnUnhandledRejection() asserts that the original unhandled
+    // rejection handler has been called for every test, so we don't need an
+    // explicit expectation here.
+    await withInstrumentOnUnhandledRejection(() => {
+      window.onunhandledrejection!({
+        reason: new Error(ERROR_MESSAGE),
+      } as PromiseRejectionEvent)
+    })
   })
 
-  it('should notify unhandled rejection', () => {
+  it('should notify unhandled rejection', async () => {
     const reason = new Error(ERROR_MESSAGE)
-    window.onunhandledrejection!({ reason } as PromiseRejectionEvent)
-
-    const [stack, originalError] = callbackSpy.calls.mostRecent().args
+    const spy = await withInstrumentOnUnhandledRejection(() => {
+      window.onunhandledrejection!({ reason } as PromiseRejectionEvent)
+    })
 
+    const [stack, originalError] = spy.calls.mostRecent().args
     expect(originalError).toBe(reason)
     expect(stack).toBeDefined()
   })
 })
 
-function setupOnUnhandledrejectionSpy() {
-  const onUnhandledrejectionSpy = jasmine.createSpy()
-  const originalOnUnhandledRejectionHandler: Window['onunhandledrejection'] = window.onunhandledrejection
-
-  if ('onunhandledrejection' in window) {
-    window.onunhandledrejection = onUnhandledrejectionSpy
-
-    registerCleanupTask(() => {
-      window.onunhandledrejection = originalOnUnhandledRejectionHandler
+/**
+ * Invokes the given callback, which is expected to be a function that throws or generates
+ * an unhandled promise rejection, and returns a promise that resolves after the callback
+ * has finished running and any global error handlers that run as a result are complete.
+ */
+function invokeAndWaitForErrorHandlers(callback: () => void): Promise<void> {
```

Comment:
I had a bug in my original implementation of this code; it lacked the nested `setTimeout`. That caused test failures on Safari in particular. Originally the buggy code was repeated in a few places in this file; I've pulled it into this helper function, fixed the issue, and added some comments to clarify why the function works in the way that it does.